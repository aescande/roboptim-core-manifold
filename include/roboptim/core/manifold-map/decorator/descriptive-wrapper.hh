#ifndef ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_DESCRIPTIVE_WRAPPER_HH
# define ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_DESCRIPTIVE_WRAPPER_HH
# include <vector>
# include <ostream>
# include <boost/shared_ptr.hpp>
# include <boost/make_shared.hpp>

# include <roboptim/core/differentiable-function.hh>

# include <manifolds/Manifold.h>

namespace roboptim
{
  /// \addtogroup roboptim_decorator
  /// @{

  // FIXME: private constructor MAGIC

  /// \brief apply a given roboptim function to a given dimension space
  /// \tparam U input function type.
  template <typename U, typename V>
  class DescriptiveWrapper
  {
  public:
    typedef boost::shared_ptr<DescriptiveWrapper> DescriptiveWrapperShPtr_t;

    /// \brief map the function on the given manifold.
    /// \param fct input function.
    /// \param functionManifold the manifold describing the function's input vector.
    template<class ... Types>
    explicit DescriptiveWrapper (Types ... args);

    DescriptiveWrapper (const U* f, pgs::Manifold& m);

    ~DescriptiveWrapper ();

    pgs::Manifold& manifold () const
    {
      return *manifold_;
    }

    const U& fct ()
    {
      return *fct_;
    }

  private:

    const U*              fct_;
    pgs::Manifold*  manifold_;
  };

  template <typename U, typename V>
  boost::shared_ptr<DescriptiveWrapper<U, V> >
  descriptivewrapper (U* fct,
		      pgs::Manifold& manifold)
  {
    return boost::make_shared<DescriptiveWrapper<U, V> > (fct, manifold);
  }

  template <typename U, typename V>
  std::ostream&
  operator<<(std::ostream& o, DescriptiveWrapper<U, V>& descWrap)
  {
    o <<
    "Displaying info about function in DescriptiveWrapper :" << "\n" <<
    "the function is " << descWrap.fct().getName() << "\n" <<
    "the function input size is " << descWrap.fct().inputSize() << "\n" <<
    "the function output size is " << descWrap.fct().outputSize() << "\n" <<
    "Displaying info about the function manifold in DescriptiveWrapper :" << "\n" <<
    "the manifold is " << descWrap.manifold().name() << "\n" <<
    "Is the manifold elementary ? : " <<
    (descWrap.manifold().isElementary() ? "yes" : "no" ) << "\n" <<
    "the manifold dimension is " << descWrap.manifold().representationDim() << "\n";
    return o;
  }

} // end of namespace roboptim.

# include <roboptim/core/manifold-map/decorator/descriptive-wrapper.hxx>
#endif //! ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_DESCRIPTIVE_WRAPPER_HH
