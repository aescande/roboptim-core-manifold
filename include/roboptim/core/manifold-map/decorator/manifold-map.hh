// Copyright (C) 2015 by Grégoire Duchemin, AIST, CNRS, EPITA
//                       Félix Darricau, AIST, CNRS, EPITA
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_MANIFOLD_MAP_HH
# define ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_MANIFOLD_MAP_HH

# include <roboptim/core/manifold-map/decorator/descriptive-wrapper.hh>
# include <roboptim/core/manifold-map/decorator/manifold-desc.hh>
# include <roboptim/core/manifold-map/decorator/function-on-manifold.hh>

#include <manifolds/SO3.h>
#include <manifolds/RealSpace.h>
#include <manifolds/CartesianProduct.h>
#include <manifolds/ExpMapMatrix.h>
#include <manifolds/S2.h>

#define ROBOPTIM_DESCRIPTIVE_FWD_DECS(NAME) typedef NAME type

#define REAL_SPACE(num) roboptim::Real<num>::Space
#define DESC_MANIFOLD(name, ...) typedef roboptim::ManiDesc< __VA_ARGS__> name
#define DEFINE_MANIFOLD_FROM_FUNCTION(name) template<class U>\
  struct Manifold_##name{\
  static pgs::Manifold* getInstance(U* function);\
  };\
  template<class U>\
  pgs::Manifold* Manifold_##name <U>::getInstance(U* function)
#define DEFINE_MANIFOLD(name) template<class U>\
  struct Manifold_##name{\
  static pgs::Manifold* getInstance(U* function);\
  };\
  template<class U>\
  pgs::Manifold* Manifold_##name <U>::getInstance(U*)

#define BIND_FUNCTION_ON_MANIFOLD(function, manifold) typedef roboptim::DescriptiveWrapper<function, manifold> function##_On_##manifold; \
  typedef roboptim::FunctionOnManifold<typename function::parent_t> Instance_##function##_On_##manifold
#define NAMED_FUNCTION_BINDING(name, function, manifold) typedef roboptim::DescriptiveWrapper<function, manifold> name; \
  typedef roboptim::FunctionOnManifold<typename function::parent_t> Instance_##name

// Library-defined elementary descriptive manifolds
// I do not think we should put those in a namespace of their own,
// because the user has to use the classes' names when (s)he defines
// a manifold.
//
// FIXME: define all remaining manifolds

namespace roboptim {

  template <class FI>
  struct S2
  {
    static pgs::Manifold* getInstance(FI*)
    {
      return new pgs::S2();
    }
  };

  template <class FI>
  struct SO3
  {
    static pgs::Manifold* getInstance(FI*)
    {
      return new pgs::SO3<pgs::ExpMapMatrix>();
    }
  };

  template<int I = 1>
  struct Real
  {
    template <class FI>
    struct Space
    {
      static pgs::Manifold* getInstance(FI*)
      {
	return new pgs::RealSpace(I);
      }
    };
  };

  template<class FI>
  struct Automated_Real
  {
  public:
    static pgs::Manifold* getInstance(FI* function)
    {
      return new pgs::RealSpace(function->getSize());
    }
  };

}

#endif //! ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_MANIFOLD_MAP_HH
