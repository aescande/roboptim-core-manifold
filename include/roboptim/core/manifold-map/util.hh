// Copyright (C) 2015 by Benjamin Chr√©tien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_MANIFOLD_MAP_UTIL_HH
# define ROBOPTIM_CORE_MANIFOLD_MAP_UTIL_HH

# include <memory>

# include <boost/shared_ptr.hpp>

namespace roboptim
{
  /// \brief Dummy deleter doing nothing.
  /// This can be used (with great care) to wrap references into smart
  /// pointers, and should be removed as soon as memory is properly handled
  /// in this project...
  /// \tparam U object type.
  template<typename U>
  struct NoopDeleter
  {
    inline void operator() (const U*) const {}
  };

  /// \brief Convert a Boost shared pointer to a STL shared pointer.
  ///
  /// Note: does not work with weak references to the source pointer.
  /// See: http://stackoverflow.com/a/12315035/1043187
  ///
  /// \tparam T object type.
  /// \param ptr Boost shared pointer.
  /// \return STL shared pointer.
  template <typename T>
  std::shared_ptr<T> toStd (boost::shared_ptr<T>& ptr)
  {
    return std::shared_ptr<T>
      (ptr.get(), [ptr](T*) mutable { ptr.reset(); });
  }

  /// \brief Convert a STL shared pointer to a Boost shared pointer.
  ///
  /// Note: does not work with weak references to the source pointer.
  /// See: http://stackoverflow.com/a/12315035/1043187
  ///
  /// \tparam T object type.
  /// \param ptr STL shared pointer.
  /// \return Boost shared pointer.
  template <typename T>
  boost::shared_ptr<T> toBoost (std::shared_ptr<T>& ptr)
  {
    return boost::shared_ptr<T>
      (ptr.get(), [ptr](T*) mutable { ptr.reset(); });
  }
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_MANIFOLD_MAP_UTIL_HH
