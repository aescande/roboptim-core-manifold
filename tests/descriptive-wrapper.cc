#include "shared-tests/fixture.hh"

#include <roboptim/core/differentiable-function.hh>

#include <roboptim/core/manifold-map/decorator/descriptive-wrapper.hh>
#include <roboptim/core/manifold-map/decorator/manifold-map.hh>

#include <manifolds/SO3.h>
#include <manifolds/RealSpace.h>
#include <manifolds/CartesianProduct.h>
#include <manifolds/ExpMapMatrix.h>
#include <manifolds/S2.h>

typedef boost::mpl::list< ::roboptim::EigenMatrixDense,
			  ::roboptim::EigenMatrixSparse> functionTypes_t;

template <class T>
struct F : public roboptim::GenericDifferentiableFunction<T>
{
  ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_(
      roboptim::GenericDifferentiableFunction<T>);

  F() : roboptim::GenericDifferentiableFunction<T>(7, 4, "dummy function") {}

  void impl_compute(result_ref, const_argument_ref) const {}

  void impl_gradient(gradient_ref, const_argument_ref, size_type) const {}
};

boost::shared_ptr<boost::test_tools::output_test_stream> output;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE_TEMPLATE (descWrapTest, T, functionTypes_t)
{
  output = retrievePattern("descriptive-wrapper");

  ROBOPTIM_DESC_MANIFOLD(Real7, ROBOPTIM_REAL_SPACE(7));
  ROBOPTIM_DESC_MANIFOLD(Real8, ROBOPTIM_REAL_SPACE(8));
  typedef roboptim::DescriptiveWrapper<F<T>, Real7> FonR7;
  typedef roboptim::DescriptiveWrapper<F<T>, Real8> FonR8;

  {
    (*output) << "DescriptiveWrapper (U* f, const mnf::Manifold& m)"
              << std::endl;
    mnf::RealSpace R7(7);
    mnf::RealSpace R8(8);
    const mnf::Manifold& mR7 = R7; 
    const mnf::Manifold& mR8 = R8; 
    const F<T>* f = new F<T>();
    FonR7 b(f, mR7);
    (*output) << b << std::endl;
    BOOST_CHECK_THROW (FonR7 c(f, mR8), std::runtime_error);   ;
  }
  {
    (*output) << "DescriptiveWrapper (const U* f, const mnf::Manifold& m)"
              << std::endl;
    mnf::RealSpace R7(7);
    mnf::RealSpace R8(8);
    const mnf::Manifold& mR7 = R7; 
    const mnf::Manifold& mR8 = R8; 
    F<T>* f = new F<T>();
    FonR7 b(f, mR7);
    (*output) << b << std::endl;
    BOOST_CHECK_THROW (FonR7 c(f, mR8), std::runtime_error);   ;
  }
  {
    (*output) << "DescriptiveWrapper (std::shared_ptr<const U> f, "
                 "std::shared_ptr<const mnf::Manifold> m)" << std::endl;
    std::shared_ptr<const mnf::Manifold> R7(new mnf::RealSpace(7));
    std::shared_ptr<const mnf::Manifold> R8(new mnf::RealSpace(8));
    std::shared_ptr<const F<T>> f(new F<T>());
    FonR7 b(f, R7);
    (*output) << b << std::endl;
    BOOST_CHECK_THROW (FonR7 b(f, R8), std::runtime_error);   ;
  }
  {
    (*output) << "DescriptiveWrapper (boost::shared_ptr<const U> f, "
                 "boost::shared_ptr<const mnf::Manifold> m)" << std::endl;
    boost::shared_ptr<const mnf::Manifold> R7(new mnf::RealSpace(7));
    boost::shared_ptr<const mnf::Manifold> R8(new mnf::RealSpace(8));
    boost::shared_ptr<const F<T>> f(new F<T>());
    FonR7 b(f, R7);
    (*output) << b << std::endl;
    BOOST_CHECK_THROW (FonR7 b(f, R8), std::runtime_error);   ;
  }
  {
    (*output) << "explicit DescriptiveWrapper (Types ... args)" << std::endl;
    FonR7 a;
    (*output) << a << std::endl;
  }
  {
    (*output) << "makeUNCHECKEDDescriptiveWrapper(const U* fct, const "
                 "mnf::Manifold& manifold)" << std::endl;
    mnf::RealSpace R7(7);
    mnf::RealSpace R8(8);
    F<T>* f = new F<T>();
    auto b = FonR7::makeUNCHECKEDDescriptiveWrapper(f, R7);
    (*output) << *b << std::endl;
    auto c = FonR8::makeUNCHECKEDDescriptiveWrapper(f, R8);
    (*output) << *c << std::endl;
  }
  {
    (*output) << "makeUNCHECKEDDescriptiveWrapper(std::shared_ptr<const U> "
                 "fct, std::shared_ptr<const mnf::Manifold> manifold)"
              << std::endl;
    boost::shared_ptr<mnf::Manifold> R7(new mnf::RealSpace(7));
    boost::shared_ptr<mnf::Manifold> R8(new mnf::RealSpace(8));
    boost::shared_ptr<F<T>> f(new F<T>());
    auto b = FonR7::makeUNCHECKEDDescriptiveWrapper(f, R7);
    (*output) << *b << std::endl;
    auto c = FonR8::makeUNCHECKEDDescriptiveWrapper(f, R8);
    (*output) << *c << std::endl;
  }
  {
    (*output) << "makeUNCHECKEDDescriptiveWrapper( boost::shared_ptr<const U> "
                 "fct, boost::shared_ptr<const mnf::Manifold> manifold)"
              << std::endl;
    std::shared_ptr<mnf::Manifold> R7(new mnf::RealSpace(7));
    std::shared_ptr<mnf::Manifold> R8(new mnf::RealSpace(8));
    std::shared_ptr<F<T>> f(new F<T>());
    auto b = FonR7::makeUNCHECKEDDescriptiveWrapper(f, R7);
    (*output) << *b << std::endl;
    auto c = FonR8::makeUNCHECKEDDescriptiveWrapper(f, R8);
    (*output) << *c << std::endl;
  }
  //std::cout << output->str() << std::endl;
  BOOST_CHECK (output->match_pattern());
}

BOOST_AUTO_TEST_SUITE_END ()
