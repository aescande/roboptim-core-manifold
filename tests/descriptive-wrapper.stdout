DescriptiveWrapper (U* f, const mnf::Manifold& m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper (const U* f, const mnf::Manifold& m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper (std::shared_ptr<const U> f, std::shared_ptr<const mnf::Manifold> m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper (boost::shared_ptr<const U> f, boost::shared_ptr<const mnf::Manifold> m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
explicit DescriptiveWrapper (Types ... args)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
makeUNCHECKEDDescriptiveWrapper(const U* fct, const mnf::Manifold& manifold)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R8
  Elementary: yes
  Dimension: 8
makeUNCHECKEDDescriptiveWrapper(std::shared_ptr<const U> fct, std::shared_ptr<const mnf::Manifold> manifold)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R8
  Elementary: yes
  Dimension: 8
makeUNCHECKEDDescriptiveWrapper( boost::shared_ptr<const U> fct, boost::shared_ptr<const mnf::Manifold> manifold)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R8
  Elementary: yes
  Dimension: 8

DescriptiveWrapper (U* f, const mnf::Manifold& m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper (const U* f, const mnf::Manifold& m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper (std::shared_ptr<const U> f, std::shared_ptr<const mnf::Manifold> m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper (boost::shared_ptr<const U> f, boost::shared_ptr<const mnf::Manifold> m)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
explicit DescriptiveWrapper (Types ... args)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
makeUNCHECKEDDescriptiveWrapper(const U* fct, const mnf::Manifold& manifold)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R8
  Elementary: yes
  Dimension: 8
makeUNCHECKEDDescriptiveWrapper(std::shared_ptr<const U> fct, std::shared_ptr<const mnf::Manifold> manifold)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R8
  Elementary: yes
  Dimension: 8
makeUNCHECKEDDescriptiveWrapper( boost::shared_ptr<const U> fct, boost::shared_ptr<const mnf::Manifold> manifold)
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R7
  Elementary: yes
  Dimension: 7
DescriptiveWrapper's function:
  Name: dummy function
  Input size: 7
  Output size: 4
DescriptiveWrapper's manifold:
  Name: R8
  Elementary: yes
  Dimension: 8

